## 探索电影数据集

在这个项目中，你将尝试使用所学的知识，使用 `NumPy`、`Pandas`、`matplotlib`、`seaborn` 库中的函数，来对电影数据集进行探索。

下载数据集：
[TMDb电影数据](https://s3.cn-north-1.amazonaws.com.cn/static-documents/nd101/explore+dataset/tmdb-movies.csv)



数据集各列名称的含义：
<table>
<thead><tr><th>列名称</th><th>id</th><th>imdb_id</th><th>popularity</th><th>budget</th><th>revenue</th><th>original_title</th><th>cast</th><th>homepage</th><th>director</th><th>tagline</th><th>keywords</th><th>overview</th><th>runtime</th><th>genres</th><th>production_companies</th><th>release_date</th><th>vote_count</th><th>vote_average</th><th>release_year</th><th>budget_adj</th><th>revenue_adj</th></tr></thead><tbody>
 <tr><td>含义</td><td>编号</td><td>IMDB 编号</td><td>知名度</td><td>预算</td><td>票房</td><td>名称</td><td>主演</td><td>网站</td><td>导演</td><td>宣传词</td><td>关键词</td><td>简介</td><td>时常</td><td>类别</td><td>发行公司</td><td>发行日期</td><td>投票总数</td><td>投票均值</td><td>发行年份</td><td>预算（调整后）</td><td>票房（调整后）</td></tr>
</tbody></table>


**请注意，你需要提交该报告导出的 `.html`、`.ipynb` 以及 `.py` 文件。**



---

---

## 第一节 数据的导入与处理

在这一部分，你需要编写代码，使用 Pandas 读取数据，并进行预处理。


**任务1.1：** 导入库以及数据

1. 载入需要的库 `NumPy`、`Pandas`、`matplotlib`、`seaborn`。
2. 利用 `Pandas` 库，读取 `tmdb-movies.csv` 中的数据，保存为 `movie_data`。

提示：记得使用 notebook 中的魔法指令 `%matplotlib inline`，否则会导致你接下来无法打印出图像。

In [31]:
import numpy as np
import pandas as pd
import matplotlib as mp
import seaborn as sb
%matplotlib inline
movie_data = pd.read_csv('./tmdb-movies.csv')
X = np.arange(25).reshape(5, 5)



---

**任务1.2: ** 了解数据

你会接触到各种各样的数据表，因此在读取之后，我们有必要通过一些简单的方法，来了解我们数据表是什么样子的。

1. 获取数据表的行列，并打印。
2. 使用 `.head()`、`.tail()`、`.sample()` 方法，观察、了解数据表的情况。
3. 使用 `.dtypes` 属性，来查看各列数据的数据类型。
4. 使用 `isnull()` 配合 `.any()` 等方法，来查看各列是否存在空值。
5. 使用 `.describe()` 方法，看看数据表中数值型的数据是怎么分布的。



In [32]:
movie_data.head()
movie_data.isnull()
movie_data.describe()
movie_data[50:61]

id    imdb_id  popularity     budget    revenue   original_title  \
50  258480  tt2402927    2.883233   11800000   40272135            Carol   
51  257211  tt2361509    2.814802   35000000  194564672       The Intern   
52  253412  tt2719848    2.798017   55000000  203427584          Everest   
53  274479  tt2446980    2.793297   60000000  101134059              Joy   
54  257088  tt2058673    2.614499  105000000  133718711      Point Break   
55  295964  tt2503944    2.584264   20000000   36606743            Burnt   
56  238615  tt2140379    2.578919   26000000   30523226        Self/less   
57  210860  tt3045616    2.575711   60000000   30418560        Mortdecai   
58  306819  tt0810819    2.557859   15000000   64191523  The Danish Girl   
59  201088  tt2717822    2.550747   70000000   17752940         Blackhat   
60  273481  tt3397884    2.487849   30000000   84025816          Sicario   

                                                 cast  \
50  Cate Blanchett|Rooney Mara|Kyle Chandler|Sarah...   
51  Robert De Niro|Anne Hathaway|Rene Russo|Anders...   
52  Jason Clarke|Jake Gyllenhaal|Josh Brolin|John ...   
53  Jennifer Lawrence|Bradley Cooper|Robert De Nir...   
54  Edgar Ramírez|Luke Bracey|Teresa Palmer|Delro...   
55  Bradley Cooper|Sienna Miller|Lily James|Alicia...   
56  Ryan Reynolds|Ben Kingsley|Natalie Martinez|Ma...   
57  Johnny Depp|Gwyneth Paltrow|Ewan McGregor|Paul...   
58  Eddie Redmayne|Alicia Vikander|Matthias Schoen...   
59  Chris Hemsworth|Leehom Wang|Tang Wei|Viola Dav...   
60  Emily Blunt|Benicio del Toro|Josh Brolin|Victo...   

                                    homepage            director  \
50                     http://carolfilm.com/         Todd Haynes   
51            http://www.theinternmovie.com/        Nancy Meyers   
52              http://www.everestmovie.com/  Baltasar Kormákur   
53       http://www.foxmovies.com/movies/joy    David O. Russell   
54  https://www.facebook.com/PointBreakMovie        Ericson Core   
55                    http://burntmovie.com/          John Wells   
56          http://selflessmovie.tumblr.com/        Tarsem Singh   
57             http://mortdecaithemovie.com/         David Koepp   
58                                       NaN          Tom Hooper   
59   http://www.legendary.com/film/blackhat/        Michael Mann   
60                                       NaN    Denis Villeneuve   

                                              tagline      ...       \
50              Some people change your life forever.      ...        
51                          Experience never gets old      ...        
52                                  The Storm Awaits.      ...        
53                                                NaN      ...        
54               The only law that matters is gravity      ...        
55  Never underestimate a man with everything to l...      ...        
56          God Created Man. Man Created Immortality.      ...        
57                         Sophistication Has a Name.      ...        
58                   Find the courage to be yourself.      ...        
59                       We are no longer in control.      ...        
60          The border is just another line to cross.      ...        

                                             overview runtime  \
50  In 1950s New York, a department-store clerk wh...     118   
51  70-year-old widower Ben Whittaker has discover...     121   
52  Inspired by the incredible events surrounding ...     121   
53  A story based on the life of a struggling Long...     124   
54  A young undercover FBI agent infiltrates a gan...     114   
55  Adam Jones is a Chef who destroyed his career ...     100   
56  An extremely wealthy elderly man dying from ca...     116   
57  Art dealer, Charles Mortdecai, searches for a ...     106   
58  When Gerda Wegener asks her husband Einar to f...     120   
59  A man is released from prison to help American...     133   
60  A young female FBI ag

---

**任务1.3: ** 清理数据

在真实的工作场景中，数据处理往往是最为费时费力的环节。但是幸运的是，我们提供给大家的 tmdb 数据集非常的「干净」，不需要大家做特别多的数据清洗以及处理工作。在这一步中，你的核心的工作主要是对数据表中的空值进行处理。你可以使用 `.fillna()` 来填补空值，当然也可以使用 `.dropna()` 来丢弃数据表中包含空值的某些行或者列。

任务：使用适当的方法来清理空值，并将得到的数据保存。

In [34]:
movie_data.fillna(method="ffill",axis=0,inplace=True)

---

---

## 第二节 根据指定要求读取数据


相比 Excel 等数据分析软件，Pandas 的一大特长在于，能够轻松地基于复杂的逻辑选择合适的数据。因此，如何根据指定的要求，从数据表当获取适当的数据，是使用 Pandas 中非常重要的技能，也是本节重点考察大家的内容。



---

**任务2.1: ** 简单读取

1. 读取数据表中名为 `id`、`popularity`、`budget`、`runtime`、`vote_average` 列的数据。
2. 读取数据表中前1～20行以及48、49行的数据。
3. 读取数据表中第50～60行的 `popularity` 那一列的数据。

要求：每一个语句只能用一行代码实现。

In [4]:
data1 = movie_data[['id','popularity','budget','runtime','vote_average']]
data2 = movie_data[:21].append(movie_data[48:50])
data3 = movie_data[50:61]['popularity']

---

**任务2.2: **逻辑读取（Logical Indexing）

1. 读取数据表中 **`popularity` 大于5** 的所有数据。
2. 读取数据表中 **`popularity` 大于5** 的所有数据且**发行年份在1996年之后**的所有数据。

提示：Pandas 中的逻辑运算符如 `&`、`|`，分别代表`且`以及`或`。

要求：请使用 Logical Indexing实现。

In [35]:
movie_data.loc[movie_data['popularity']>5]
movie_data.loc[movie_data['popularity']>5].loc[movie_data['release_year']>1996]

id    imdb_id  popularity     budget     revenue  \
0     135397  tt0369610   32.985763  150000000  1513528810   
1      76341  tt1392190   28.419936  150000000   378436354   
2     262500  tt2908446   13.112507  110000000   295238201   
3     140607  tt2488496   11.173104  200000000  2068178225   
4     168259  tt2820852    9.335014  190000000  1506249360   
5     281957  tt1663202    9.110700  135000000   532950503   
6      87101  tt1340138    8.654359  155000000   440603537   
7     286217  tt3659388    7.667400  108000000   595380321   
8     211672  tt2293640    7.404165   74000000  1156730962   
9     150540  tt2096673    6.326804  175000000   853708609   
10    206647  tt2379713    6.200282  245000000   880674609   
11     76757  tt1617661    6.189369  176000003   183987723   
12    264660  tt0470752    6.118847   15000000    36869414   
13    257344  tt2120120    5.984995   88000000   243637091   
14     99861  tt2395427    5.944927  280000000  1405035767   
15    273248  tt3460252    5.898400   44000000   155760117   
16    260346  tt2446042    5.749758   48000000   325771424   
17    102899  tt0478970    5.573184  130000000   518602163   
18    150689  tt1661199    5.556818   95000000   542351353   
19    131634  tt1951266    5.476958  160000000   650523427   
20    158852  tt1964418    5.462138  190000000   209035668   
21    307081  tt1798684    5.337064   30000000    91709827   
629   157336  tt0816692   24.949134  165000000   621752480   
630   118340  tt2015381   14.311205  170000000   773312399   
631   100402  tt1843866   12.971027  170000000   714766572   
632   245891  tt2911666   11.422751   20000000    78739897   
633   131631  tt1951265   10.739009  125000000   752100229   
634   122917  tt2310332   10.174599  250000000   955119788   
635   177572  tt2245084    8.691294  165000000   652105443   
636   205596  tt2084970    8.110711   14000000   233555708   
...      ...        ...         ...        ...         ...   
1922   44214  tt0947798    5.293180   13000000   327803731   
2409     550  tt0137523    8.947905   63000000   100853753   
2410     603  tt0133093    7.753899   63000000   463517383   
2633     120  tt0120737    8.575419   93000000   871368364   
2634     671  tt0241527    8.021423  125000000   976475550   
2875     155  tt0468569    8.466668  185000000  1001921825   
2876   10681  tt0910970    5.678119  180000000   521311860   
3371  161337  tt2381375    8.411577          0           0   
3372    1771  tt0458339    7.959228  140000000   370569774   
3373   64690  tt0780504    5.903353   15000000    76175166   
3374   12445  tt1201607    5.711315  125000000  1327817822   
3911     121  tt0167261    8.095275   79000000   926287400   
3912     672  tt0295297    6.012584  100000000   876688482   
4361   24428  tt0848228    7.637767  220000000  1519557910   
4362   52520  tt1496025    7.031452   70000000   132400000   
4363   49026  tt1345836    6.591277  250000000  1081041287   
4364   68718  tt1853728    5.944518  100000000   425368238   
4365   37724  tt1074638    5.603587  200000000  1108561013   
4949     122  tt0167260    7.122455   94000000  1118888979   
4950     277  tt0320691    6.887883   22000000    95708457   
4951      22  tt0325980    6.864067  140000000   655011224   
4952      24  tt0266697    6.174132   30000000   180949000   
5230   13590  tt0195753    6.668990          0           0   
5422  109445  tt2294629    6.112766  150000000  1274219009   
5423   49047  tt1454468    5.242753  105000000   716392705   
5424   76338  tt1981115    5.111900  170000000   479765000   
6190     674  tt0330373    5.939927  150000000   895921036   
6191     272  tt0372784    5.400826  150000000   374218673   
6554     834  tt0401855    5.838503   50000000   111340801   
6962     673  tt0304141    5.827781  130000000   789804554   

                                         original_title  \
0                                        Jurassic World   
1                                    Mad Max: F

---

**任务2.3: **分组读取

1. 对 `release_year` 进行分组，使用 [`.agg`](http://pandas.pydata.org/pandas-docs/version/0.22/generated/pandas.core.groupby.DataFrameGroupBy.agg.html) 获得 `revenue` 的均值。
2. 对 `director` 进行分组，使用 [`.agg`](http://pandas.pydata.org/pandas-docs/version/0.22/generated/pandas.core.groupby.DataFrameGroupBy.agg.html) 获得 `popularity` 的均值，从高到低排列。

要求：使用 `Groupby` 命令实现。

In [36]:
movie_data.groupby('release_year').agg('mean')['revenue']

release_year
1960    4.531406e+06
1961    1.089420e+07
1962    6.736870e+06
1963    5.511911e+06
1964    8.118614e+06
1965    1.308805e+07
1966    1.842102e+06
1967    1.844587e+07
1968    6.788025e+06
1969    7.869583e+06
1970    1.366395e+07
1971    7.362011e+06
1972    1.236825e+07
1973    2.225420e+07
1974    1.728808e+07
1975    2.176114e+07
1976    1.704267e+07
1977    3.825584e+07
1978    2.107353e+07
1979    2.955781e+07
1980    2.267516e+07
1981    2.164154e+07
1982    3.035116e+07
1983    2.884412e+07
1984    2.510023e+07
1985    2.638323e+07
1986    2.481635e+07
1987    2.769684e+07
1988    2.579001e+07
1989    3.770017e+07
1990    4.026641e+07
1991    3.538797e+07
1992    4.570040e+07
1993    3.907388e+07
1994    3.856212e+07
1995    5.232195e+07
1996    4.074261e+07
1997    5.549569e+07
1998    4.520559e+07
1999    5.069515e+07
2000    4.836432e+07
2001    5.541357e+07
2002    5.505120e+07
2003    5.387275e+07
2004    5.470301e+07
2005    4.537592e+07
2006    3.989152e+07


---

---

## 第三节 绘图与可视化

接着你要尝试对你的数据进行图像的绘制以及可视化。这一节最重要的是，你能够选择合适的图像，对特定的可视化目标进行可视化。所谓可视化的目标，是你希望从可视化的过程中，观察到怎样的信息以及变化。例如，观察票房随着时间的变化、哪个导演最受欢迎等。

<table>
<thead><tr><th>可视化的目标</th><th>可以使用的图像</th></tr></thead><tbody>
 <tr><td>表示某一属性数据的分布</td><td>饼图、直方图、散点图</td></tr>
 <tr><td>表示某一属性数据随着某一个变量变化</td><td>条形图、折线图、热力图</td></tr>
 <tr><td>比较多个属性的数据之间的关系</td><td>散点图、小提琴图、堆积条形图、堆积折线图</td></tr>
</tbody></table>

在这个部分，你需要根据题目中问题，选择适当的可视化图像进行绘制，并进行相应的分析。对于选做题，他们具有一定的难度，你可以尝试挑战一下～

**任务3.1：**对 `popularity` 最高的20名电影绘制其 `popularity` 值。

---
**任务3.2：**分析电影净利润（票房-成本）随着年份变化的情况，并简单进行分析。

---

**[选做]任务3.3：**选择最多产的10位导演（电影数量最多的），绘制他们排行前3的三部电影的票房情况，并简要进行分析。

---

**[选做]任务3.4：**分析1968年~2015年六月电影的数量的变化。

---

**[选做]任务3.5：**分析1968年~2015年六月电影 `Comedy` 和 `Drama` 两类电影的数量的变化。

> 注意: 当你写完了所有的代码，并且回答了所有的问题。你就可以把你的 iPython Notebook 导出成 HTML 文件。你可以在菜单栏，这样导出**File -> Download as -> HTML (.html)、Python (.py)** 把导出的 HTML、python文件 和这个 iPython notebook 一起提交给审阅者。